### DISCUSSION TITLE
Questions about USB vendor specific interface

### DISCUSSION THREAD

**Discussion by sdbbs** [hathach/tinyusb#1619] [General]
I am trying to study the USB vendor specific interface. I work on a Pico/RP2040 board, and I noticed there is a so-called "Reset" interface which is vendor specific, which is shown under 'Other devices' in Windows 10 Device Manager:

![image](https://user-images.githubusercontent.com/48317456/187187547-0166634a-4578-43dd-b414-64ba719676ab.png)

As I understand it, the "orange" exclamation mark is there because this device does not have a driver loaded - which, I assume, is OK, considering this is a "vendor" interface.

Before I proceed, let me note that I've found "Vendor Descriptor Templates" in [tinyusb/usbd.h](https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h):

```c
// Interface number, string index, EP Out & IN address, EP size
#define TUD_VENDOR_DESCRIPTOR(_itfnum, _stridx, _epout, _epin, _epsize) \
  /* Interface */\
  9, TUSB_DESC_INTERFACE, _itfnum, 0, 2, TUSB_CLASS_VENDOR_SPECIFIC, 0x00, 0x00, _stridx,\
  /* Endpoint Out */\
  7, TUSB_DESC_ENDPOINT, _epout, TUSB_XFER_BULK, U16_TO_U8S_LE(_epsize), 0,\
  /* Endpoint In */\
  7, TUSB_DESC_ENDPOINT, _epin, TUSB_XFER_BULK, U16_TO_U8S_LE(_epsize), 0
```

I read this as: the interface numbered `_itfnum` has 0 for alternative setting value, and 2 endpoints in total, is of `VENDOR_SPECIFIC` class with subclass 0x00 and protocol 0x00 and string description at index `_stridx` of whatever array holds the descriptor strings; and the 2 endpoints are: output bulk endpoint, and input bulk endpoint.

So, I found that the Pico Reset interface is defined in [pico-sdk/reset_interface.c](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_stdio_usb/reset_interface.c); so I copied that into a separate project, however I did not want any CDC - I wanted *only* the vendor-specific interface, so I modified the descriptor from [pico-sdk/stdio_usb_descriptors.c](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_stdio_usb/stdio_usb_descriptors.c) to be just:

```c
#define USBD_DESC_LEN (TUD_CONFIG_DESC_LEN + TUD_RPI_RESET_DESC_LEN)
#define USBD_ITF_RPI_CUSTOM (0)
#define USBD_ITF_MAX       (1)
// ...
#define TUD_RPI_RESET_DESCRIPTOR(_itfnum, _stridx) \
  /* Interface */\
  9, TUSB_DESC_INTERFACE, _itfnum, 0, 0, TUSB_CLASS_VENDOR_SPECIFIC, RESET_INTERFACE_SUBCLASS, RESET_INTERFACE_PROTOCOL, _stridx,

static const uint8_t usbd_desc_cfg[USBD_DESC_LEN] = {
    TUD_CONFIG_DESCRIPTOR(1, USBD_ITF_MAX, USBD_STR_0, USBD_DESC_LEN,
        0, USBD_MAX_POWER_MA),

    TUD_RPI_RESET_DESCRIPTOR(USBD_ITF_RPI_RESET, USBD_STR_RPI_RESET)
};
```

Note that there are no endpoints defined here anymore (as in `TUD_VENDOR_DESCRIPTOR`) - but I did not think it would be a problem, because the functionality in [pico-sdk/reset_interface.c](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_stdio_usb/reset_interface.c) is implemented in the "vendor interface control request" callback `resetd_control_xfer_cb`.

(note also that the *device* is not "vendor specific", but instead it is of class Miscellaneous `TUSB_CLASS_MISC`; it is only the interface that is "vendor specific", that is `TUSB_CLASS_VENDOR_SPECIFIC`)

I got this project to compile, and after flashing the Pico, I indeed get just a single device under 'Other devices'; however:

* libusb on the PC side - which I hoped to use to issue the different control codes handled in `resetd_control_xfer_cb` - cannot open this device (fails with -5 LIBUSB_ERROR_NOT_FOUND), and so cannot read string descriptors
* [usbview](https://github.com/Microsoft/Windows-driver-samples/tree/master/usb/usbview) can read string descriptors few seconds after flashing the Pico - but for the most part, it cannot read any string descriptors, failing with:

```none
Current Config Value:              0x00  -> Device Bus Speed: Full (is not SuperSpeed or higher capable)
Device Address:                    0x19
Open Pipes:                           0
*!*ERROR:  No open pipes!
...
iManufacturer:                     0x01
String Descriptor for index 1 not available while device is in low power state.
iProduct:                          0x02
String Descriptor for index 2 not available while device is in low power state.
...
```

I'm not really sure even how to confirm the power state of a USB device in Windows, since Device Manager and usbview can give contradictory information (https://stackoverflow.com/questions/73490266/retrieve-windows-usb-device-power-status-from-the-command-line), however looking at https://stackoverflow.com/questions/51025878/how-to-get-string-descriptor-from-a-usb-device-which-is-in-low-power-state

> The USB 2.0 specification defines suspend mode, a state where there is no traffic on the bus and devices go to sleep to save power. Since there is no traffic, you cannot request string descriptors from a device in suspend mode. You'd have to wake it up with special signalling before you can do that.

... my guess is, seeing there are on endpoints/"open pipes", Window puts the device to sleep, which then prevents libusb to open the device and usbview to read the string descriptors.

So, in summary, my questions are:

* If I want just a vendor interface (and nothing else), must I also define at least an input endpoint? If so - how come [pico-sdk/reset_interface.c](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_stdio_usb/reset_interface.c) doesn't define any for its vendor interface?
* Since I've noticed libusb on Windows cannot open even the "original"  (i.e. when it also has CDC endpoints) pico_stdio_usb Reset interface, but it can open Picoprobe - must I also specify a driver for Windows, to allow libusb to talk to this device? If so, is there any guide for this? (I've noticed Picoprobe uses a libwdi driver, which makes Picoprobe  a "WinUSB Generic Device")

**Comment by sdbbs** [comment]
Well, having looked into this further, I can answer:

> * If I want just a vendor interface (and nothing else), must I also define at least an input endpoint?

No - a firmware with just a config for a vendor interface can be recognized as a valid USB device by a USB host.

> * ... must I also specify a driver for Windows, to allow libusb to talk to this device? 

Yes. There are plenty of caveats, most important for me was this this:

https://github.com/WICG/webusb/issues/143#issuecomment-440348384

> The Windows driver stack does not allow any application to connect to a device at the USB level if the WinUSB or libusbk driver is not bound to the device.

So eventually libusbK worked as a driver for me (note there is a whole procedure about signing).

Note that even with no endpoints, one needs to have one interface defined, otherwise the configuration descriptor strings cannot be read.

Windows as a USB host will try to suspend the device, especially if there are exclamation marks for the device entry in Device Manager -- and this makes the config descriptor strings unreadable; so power management should be looked into on both driver and firmware side, to minimize chances of suspending the device.

**Comment by hathach** [comment]
glad you figure it out, most of the issue is inconsistent descriptor. Exclamation mark in device manager can be caused by several reasons, you should right click and try to find more detail on the cause. You could also want to check the windows/kernel log to see why windows decide to suspend usb device.

**Comment by sdbbs** [comment]
Thanks for the feedback, @hathach : 

> Exclamation mark in device manager can be caused by several reasons, you should right click and try to find more detail on the cause. 

That was a tricky one for me, because a lot of times I just got:

```none
This device cannot start. (Code 10)

{Operation Failed}
The requested operation was unsuccessful.
```

... and of course, it doesn't tell you *which* requested operation was unsuccessful `:)` 

> You could also want to check the windows/kernel log to see why windows decide to suspend usb device.

As far as I can tell, the standard "kernel" log in Windows 10 is in Event Viewer app, under Windows Logs\System.
This was impossible for me to read, as it logs *everything* kernel related, not just USB - and I cannot tell still how to filter events in Event Viewer.
At the time I didn't find a command-line Event viewer tool, but since I've found (via https://superuser.com/questions/1665144/querying-event-viewer-in-cmd) `eventquery.vbs` (deprecated) and [wevtutil](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/wevtutil) - maybe those would have made searching through the event viewer easier...

As for USB, usbpcap/Wireshark did not show all URBs I expected to see; I found instead [USB3HWVerifierAnalyzer.exe](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/how-to-retrieve-information-about-a-usb-device) - it seemingly captures *everything* on the USB bus in .etl format (which opens in Event Viewer); then you can also convert the log to text, but it is FULL of entries like

```none
...
12.278031: (UsbHub3/110)
        Event Message: State Machine Event
        PortPath:  0x6, 0x3, 0x0, 0x0, 0x0, 0x0
12.278032: (UsbHub3/110)
        Event Message: State Machine Event
        PortPath:  0x6, 0x3, 0x0, 0x0, 0x0, 0x0
...
```

... which again, tell me absolutely nothing - I could (probably) determine which device it is about via PortPath, but ... *which* "State Machine Event" occured, is impossible to know.

So, I will admit - even at this time, I cannot tell exactly *why* Windows suspended the device (as I cannot find an explicit record of the USB suspend reason anywhere); the only thing I can say from my limited experience is that - as soon as my firmware and driver setup got good enough so the USB device was shown in Device Manager "clean", these suspends stopped.

Due to this behavior, I can speculate that if Windows has your device without a driver or with an exclamation mark, it probably thinks "Ah, nothing can use this device anyways, let me suspend it so it saves power" and issues a suspend within 5-10 seconds after such a device is attached - but who knows how correct this speculation is ...

**Comment by hathach** [comment]
can you post txt file containing the tusb log with `LOG=2`. Also you could try to enumerate your device with linux to see if it is working there. Linux kernel log is easier to access with `dmesg`

**Comment by sdbbs** [comment]
Thanks @hathach : 

> can you post txt file containing the tusb log with `LOG=2`

Ah, no worries, I did not need help - I was just inspired by your earlier comment, to complain about my experience with Windows `:)`

> Also you could try to enumerate your device with linux to see if it is working there. Linux kernel log is easier to access with `dmesg`

Indeed - unfortunately my dev hardware is attached to a Windows machine, and where it is, I do not have access to a Linux machine; otherwise that is how I too would have started development!
